[FLINK-7973] disable JNI bridge for relocated hadoop classes in s3-fs-*

NicoK opened the pull request:
## What is the purpose of the change

If some Hadoop's JNI library is in the classpath, it will be loaded by our shaded, relocated hadoop classes in the `flink-s3-fs-*` filesystems as well. Then, however, `NativeCodeLoader#isNativeCodeLoaded` will return `true` and native code libraries will be tried although our relocated namespaces have no JNI mapping leading to errors like `java.lang.UnsatisfiedLinkError: org.apache.flink.fs.s3hadoop.shaded.org.apache.hadoop.security.JniBasedUnixGroupsMapping.anchorNative()V`.

## Brief change log

- disable native code loading (there are more users than the shown `JniBasedUnixGroupsMapping`) via copies of the respective `NativeCodeLoader` class

## Verifying this change

This change added tests and can be verified as follows:

  - Manually verified the change by running a 3 node cluster with 1 JobManagers and 2 TaskManagers on EMR executing the `WordCount` example with an S3 input source:
  ```
cp ./opt/flink-s3-fs-hadoop-1.4-SNAPSHOT.jar ./lib/
./bin/flink run -m yarn-cluster -yn 2 -ys 1 -yjm 768 -ytm 1024 ./examples/batch/WordCount.jar --input s3://<bucket>/<path-to-intput-file>
```

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **yes** -- actually, the shaded and relocated Hadoop classes may not use (potentially faster) JNI implementations for certain functions; depending on their use, this may be per record but since this only applies to the S3 filesystem access, performance penalties should be hidden by its access times anyway
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: **no**
  - The S3 file system connector: **yes**

## Documentation

  - Does this pull request introduce a new feature? **no**
  - If yes, how is the feature documented? **docs**

zentol commented:
Changes look reasonable to me. I supposed we can't guard this with a test?

NicoK commented:
This would be difficult to test - the most reasonable test would be to put the Hadoop JNI into the classpath and then run the end-to-end test on that but we don't want to include the platform-dependant jni library, I suppose

aljoscha commented:
Unfortunately, I don't think we can test this now. I have in mind to add more end-to-end tests or "system tests" that run nightly on AWS or some proper Hadoop and they should catch it.

What do you think, @zentol?

zentol commented:
I expected that we can't test automatically this for now. From my side, +1 to merge.

We definitely need more tests that actually run against flink-dist though.

aljoscha commented:
Jip, the end-to-end tests are a first step but we need more. 👌

aljoscha commented:
@NicoK could you please close this PR?

And thanks for the work! 👍

NicoK commented:
Fixed on master in
0e5fb0b78cd0a3ccb144071a47579eb6c3d0570a
e9e7c3372189db7e933ff59114b9ec6245838eda

Fixed on release-1.4 in
25a28ab32609c45fb8c40f717148e32fb453d2fc
9f68212603e3601e2f7a67ff93be9b15844c14da
